pkg/sanity/sanity.go

```diff
@@ -149,55 +165,66 @@ type SanityContext struct {
 	StagingPath string
 }
 
+// NewTestConfig returns a config instance with all values set to
+// their defaults.
+func NewTestConfig() TestConfig {
+	return TestConfig{
+		TargetPath:           os.TempDir() + "/csi-mount",
+		StagingPath:          os.TempDir() + "/csi-staging",
+		CreatePathCmdTimeout: 10 * time.Second,
+		RemovePathCmdTimeout: 10 * time.Second,
+		TestVolumeSize:       10 * 1024 * 1024 * 1024, // 10 GiB
+		IDGen:                &DefaultIDGenerator{},
+	}
+}
+
 // newContext sets up sanity testing with a config supplied by the
 // user of the sanity package. Ownership of that config is shared
 // between the sanity package and the caller.
-func newContext(reqConfig *Config) *SanityContext {
-	// To avoid runtime if checks when using IDGen, a default
-	// is set here.
-	if reqConfig.IDGen == nil {
-		reqConfig.IDGen = &DefaultIDGenerator{}
-	}
-
-	return &SanityContext{
-		Config: reqConfig,
+func newTestContext(config *TestConfig) *TestContext {
+	return &TestContext{
+		Config: config,
 	}
 }
 
 // Test will test the CSI driver at the specified address by
 // setting up a Ginkgo suite and running it.
-func Test(t *testing.T, reqConfig *Config) {
-	// Get StorageClass parameters from TestVolumeParametersFile
-	loadFromFile(reqConfig.TestVolumeParametersFile, &reqConfig.TestVolumeParameters)
-	// Get VolumeSnapshotClass parameters from TestSnapshotParametersFile
-	loadFromFile(reqConfig.TestSnapshotParametersFile, &reqConfig.TestSnapshotParameters)
-
-	sc := newContext(reqConfig)
-	registerTestsInGinkgo(sc)
+func Test(t *testing.T, config TestConfig) {
+	sc := GinkgoTest(&config)
 	RegisterFailHandler(Fail)
 
 	var specReporters []Reporter
-	if reqConfig.JUnitFile != "" {
-		junitReporter := reporters.NewJUnitReporter(reqConfig.JUnitFile)
+	if config.JUnitFile != "" {
+		junitReporter := reporters.NewJUnitReporter(config.JUnitFile)
 		specReporters = append(specReporters, junitReporter)
 	}
 	RunSpecsWithDefaultAndCustomReporters(t, "CSI Driver Test Suite", specReporters)
-	if sc.Conn != nil {
-		sc.Conn.Close()
-	}
+	sc.Finalize()
 }
 
-// GinkoTest is another entry point for sanity testing: instead of directly
-// running tests like Test does, it merely registers the tests. This can
-// be used to embed sanity testing in a custom Ginkgo test suite.
-func GinkgoTest(reqConfig *Config) {
-	sc := newContext(reqConfig)
+// GinkoTest is another entry point for sanity testing: instead of
+// directly running tests like Test does, it merely registers the
+// tests. This can be used to embed sanity testing in a custom Ginkgo
+// test suite.  The pointer to the configuration is merely stored by
+// GinkgoTest for use when the tests run. Therefore its content can
+// still be modified in a BeforeEach. The sanity package itself treats
+// it as read-only.
+func GinkgoTest(config *TestConfig) *TestContext {
```

Comment:
Maybe `RegisterGinkgoTests` would be a good name for this, but we seem to have enough breaking changes on our hands.